ec2/sc-ec2-linux-jumpcloud-notebook.yaml

```diff
@@ -249,6 +252,80 @@ Resources:
       Roles:
         - !Ref 'InstanceRole'
 
+  # The EC2 instance writes to it, but this is introduced for the cleanup of the parameter
+  InstanceSSMParameter:
+    Type: AWS::SSM::Parameter
+    Properties:
+      # This will need to be synced with the `packer-rstudio` project
+      Name: !Sub "/service-catalog/synapse/cred/${LinuxInstance}/odic-accesstoken"
+      Type: String
+      Value: to be replaced by EC2
+
+  SSMParameterEncryptionKey:
+    Type: AWS::KMS::Key
+    Properties:
+      Description: key used to encrypt SSM Parameters for a EC2 instance
+      # number of days to wait before key gets deleted
+      PendingWindowInDays: 7
+      KeyPolicy:
+        Version: "2012-10-17"
+        Statement:
+        - Sid: "Deny decrypt to all Principals that shouldn't require access"
+          Effect: 'Deny'
+          NotPrincipal:
+            AWS:
+              - !ImportValue
+                'Fn::Sub': '${AWS::Region}-sc-ec2vpc-launchrole-LaunchRoleArn'
+          Action:
+            - kms:Decrypt
+          Resource: '*'
+        - Sid: Allow use of the key
+          Effect: Allow
+          Principal:
+            AWS: !Sub "${InstanceRole.Arn}"
+          Action:
+          - kms:DescribeKey
+          - kms:Encrypt
+          - kms:Decrypt
+          - kms:ReEncrypt*
+          - kms:GenerateDataKey
+          - kms:GenerateDataKeyWithoutPlaintext
+          Resource: '*'
+        - Sid: 'Default KMS Policy that enables IAM permissions'  # https://docs.aws.amazon.com/kms/latest/developerguide/key-policies.html#key-policy-default
+          Effect: 'Allow'
+          Principal:
+            AWS: !Sub 'arn:aws:iam::${AWS::AccountId}:root'
+          Action:
+            - 'kms:*'
+          Resource: '*'
+        - Sid: "Allow administration of the key to CFN service role"
+          Effect: "Allow"
+          Principal:
+            AWS:
+              - !ImportValue
+                'Fn::Sub': '${AWS::Region}-sc-ec2vpc-launchrole-LaunchRoleArn'
+          Action:
+            - "kms:Create*"
+            - "kms:Describe*"
+            - "kms:Enable*"
+            - "kms:List*"
+            - "kms:Put*"
+            - "kms:Update*"
+            - "kms:Revoke*"
+            - "kms:Disable*"
+            - "kms:Get*"
+            - "kms:Delete*"
+            - "kms:ScheduleKeyDeletion"
+            - "kms:CancelKeyDeletion"
+          Resource: "*"
+
+  SSMParameterEncryptionKeyAlias:
+    Type: AWS::KMS::Alias
+    Properties:
+      # This will need to be synced with the `packer-rstudio` project
+      AliasName: !Sub 'alias/service-catalog/synapse/cred/${LinuxInstance}'
+      TargetKeyId: !Ref SSMParameterEncryptionKey
+
   LinuxInstance:
     Type: AWS::EC2::Instance
```

Comment:
is this EC2 writing the secret to the SSM parameter?  if so we need to make sure the kms key and SSM parameter exists before this gets created.  I think you can do that by adding a `DependsOn`..

```
DependsOn:
  - InstanceSSMParameter
  - SSMParameterEncryptionKey
```

Reply:
This would cause a circular dependency w/ the EC2 instance. The naming of these parameters depend on the EC2 instance id. The first time EC2 instance would write to SSM Parameter Store would be upon user interaction w/ the RStudio URL so it shouldn't happen until after everything is instantiated.